### PR TITLE
KAFKA-10555: Improve client state machine

### PR DESCRIPTION
**Pull request by wcarlson5**
Removes the transition to error for when there are no threads and makes Error terminal

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -532,22 +537,6 @@ public void setGlobalStateRestoreListener(final StateRestoreListener globalState
             this.threadStatesLock = new Object();
         }
 
-        /**
-         * If all threads are dead set to ERROR
-         */
-        private void maybeSetError() {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:558]: This will not be needed with the new error definition

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -594,9 +581,8 @@ public synchronized void onChange(final Thread thread,
                     if (newState == GlobalStreamThread.State.RUNNING) {
                         maybeSetRunning();
                     } else if (newState == GlobalStreamThread.State.DEAD) {
-                        if (setState(State.ERROR)) {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:617]: we will be doing the same thing but closing the client for now. Maybe a replace globalThread will be added later

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 3]
```diff
@@ -1142,11 +1128,21 @@ private Thread shutdownHelper(final boolean error) {
...
             if (!error) {
                 setState(State.NOT_RUNNING);
+            } else {
+                setState(State.ERROR);
             }
         }, "kafka-streams-close-thread");
     }
 
     private boolean close(final long timeoutMs) {
+        if (state == State.ERROR) {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1200]: We want to make `close()` idempotent and not throw an exception but we will log a warning, but only for close so that is why these logs are not in the `setState()` method.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -546,7 +546,7 @@ public void run() {
      * @throws IllegalStateException If store gets registered after initialized is already finished
      * @throws StreamsException      if the store's change log does not contain the partition
      */
-    void runLoop() {
+    boolean runLoop() {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:555]: This will let Streams shutdown uncleanly when in EOS mode

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 5]
```diff
@@ -418,70 +418,6 @@ public void stateShouldTransitToRunningIfNonDeadThreadsBackToRunning() throws In
         Assert.assertEquals(KafkaStreams.State.NOT_RUNNING, streams.state());
     }
 
-    @Test
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:431]: This test is for the functionality we are removing

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 6]
```diff
@@ -625,8 +561,7 @@ public void shouldNotAddThreadWhenError() {
         final KafkaStreams streams = new KafkaStreams(getBuilderWithSource().build(), props, supplier, time);
         final int oldSize = streams.threads.size();
         streams.start();
-        streamThreadOne.shutdown();
-        streamThreadTwo.shutdown();
+        globalStreamThread.shutdown();
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:592]: removing stream threads does not put the client in error anymore. The global does

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 7]
```diff
@@ -524,9 +460,6 @@ public void testStateThreadClose() throws Exception {
                     "Thread never stopped.");
                 streams.threads.get(i).join();
             }
-            TestUtils.waitForCondition(
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:537]: There is no transition to Error here anymore

File: streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java [hunk 8]
```diff
@@ -99,7 +99,6 @@ public static void setupConfigsAndUtils() {
     }
 
     @Test
-    @SuppressWarnings("deprecation")
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java:102]: remove deprecation

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 9]
```diff
@@ -411,7 +405,8 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
                 errorInjectedClient1.set(false);
                 stateTransitions1.clear();
                 streams1Alpha.close();
-                waitForStateTransition(stateTransitions1, CLOSE_CRASHED);
+                final KafkaStreams finalStreams1Alpha = streams1Alpha;
+                waitForCondition(() -> finalStreams1Alpha.state() == State.ERROR, "Stream did not go to ERROR");
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Just wait for ERROR because you don't close crashed

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 10]
```diff
@@ -526,12 +521,10 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
 
                 assignmentListener.waitForNextStableAssignment(MAX_WAIT_TIME_MS);
 
-                waitForStateTransition(stateTransitions2, CRASH);
-
                 commitErrorInjectedClient2.set(false);
                 stateTransitions2.clear();
                 streams2Alpha.close();
-                waitForStateTransition(stateTransitions2, CLOSE_CRASHED);
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: No longer do you need to close crashed

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java [hunk 11]
```diff
@@ -174,8 +177,8 @@ private KafkaStreams buildStreamWithDirtyStateDir(final String stateDirPath,
     }
 
     @Test
-    @Deprecated
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:177]: remove deprecated handler

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java [hunk 12]
```diff
@@ -174,8 +177,8 @@ private KafkaStreams buildStreamWithDirtyStateDir(final String stateDirPath,
...
     public void shouldWipeOutStandbyStateDirectoryIfCheckpointIsMissing() throws Exception {
+        final long time = System.currentTimeMillis();
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: There was a problem with the cleaner thread sometimes wiping out old segments because they were expired

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java [hunk 13]
```diff
@@ -139,7 +139,7 @@ public void shouldShutdownClient() throws InterruptedException {
             StreamsTestUtils.startKafkaStreamsAndWaitForRunningState(kafkaStreams);
 
             produceMessages(0L, inputTopic, "A");
-            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.NOT_RUNNING, DEFAULT_DURATION);
+            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.ERROR, DEFAULT_DURATION);
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java:148]: When the client is shutdown it now goes to ERROR

**Review by wcarlson5 — COMMENTED** [review]
@mjsax Can you give this a look? I hope that I haven't broken any of the EOS tests

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 14]
```diff
@@ -226,7 +226,8 @@
         RUNNING(1, 2, 3, 5),    // 2
         PENDING_SHUTDOWN(4),    // 3
         NOT_RUNNING,            // 4
-        ERROR(3);               // 5
+        PENDING_ERROR(6),       // 5
+        ERROR;               // 6
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: super nit: fix indention of comment

Also please update the state diagram in the comment above.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 15]
```diff
@@ -1165,11 +1151,21 @@ private Thread shutdownHelper(final boolean error) {
...
+            } else {
+                setState(State.ERROR);
             }
         }, "kafka-streams-close-thread");
     }
 
     private boolean close(final long timeoutMs) {
+        if (state == State.ERROR) {
+            log.info("Streams client is already in the terminal state ERROR, all resources are closed and the client has stopped.");
+            return false;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Should we return `true` there? I understand that we are not in `NOT_RUNNING` state, but in the end we are in a terminal state and we did cleanup all resources. -- I guess, I am raising the question if we should relax the definition of this return value?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 7 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:537]: Should we add a check that the thread count went down to zero? We could use the corresponding metrics.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 16]
```diff
@@ -557,12 +490,12 @@ public void testStateGlobalThreadClose() throws Exception {
...
                 "Thread never stopped.");
             globalStreamThread.join();
-            assertEquals(streams.state(), KafkaStreams.State.ERROR);
+            assertEquals(streams.state(), KafkaStreams.State.PENDING_ERROR);
         } finally {
             streams.close();
         }
 
-        assertEquals(streams.state(), KafkaStreams.State.NOT_RUNNING);
+        assertEquals(streams.state(), KafkaStreams.State.PENDING_ERROR);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: If the global thread dies, should we not transit to `ERROR` at the end?

File: streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java (hunk 8 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java:102]: Why did we even set the exception handler that does not do anything? 🤔

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 17]
```diff
@@ -388,7 +382,8 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
                 errorInjectedClient1.set(false);
                 stateTransitions1.clear();
                 streams1Alpha.close();
-                waitForStateTransition(stateTransitions1, CLOSE_CRASHED);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:402]: Why not just change this from `CLOSE_CRASHED` to `CRASH` ?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 18]
```diff
@@ -113,14 +114,7 @@
...
             Collections.singletonList(
-                KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.ERROR)
-            )
-        );
-    private static final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> CLOSE_CRASHED =
-        Collections.unmodifiableList(
-            Arrays.asList(
-                KeyValue.pair(KafkaStreams.State.ERROR, KafkaStreams.State.PENDING_SHUTDOWN),
-                KeyValue.pair(KafkaStreams.State.PENDING_SHUTDOWN, KafkaStreams.State.NOT_RUNNING)
+                KeyValue.pair(KafkaStreams.State.RUNNING, State.PENDING_ERROR)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Should we extend it to have two state transitions, `RUNNING -> PENDING_ERROR` and `PENDING_ERROR -> ERROR` ?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 19]
```diff
@@ -521,12 +516,10 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
 
                 assignmentListener.waitForNextStableAssignment(MAX_WAIT_TIME_MS);
 
-                waitForStateTransition(stateTransitions2, CRASH);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:536]: Why do we remove this? If the fail, we should transit to PENDING_ERROR state first, and only afterwards move forward with the test.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: Why this change? client2 should transit from `PENDING_ERROR` to `ERROR` (not from `PENDING_SHUTDOWN` to `CLOSED`, and thus the new `CLOSE` check seem not to be right? Or do I miss something?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 20]
```diff
@@ -611,12 +604,11 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
                 verifyUncommitted(expectedUncommittedResult);
 
                 assignmentListener.waitForNextStableAssignment(MAX_WAIT_TIME_MS);
-                waitForStateTransition(stateTransitions1, CRASH);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:627]: As above

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 21]
```diff
@@ -611,12 +604,11 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
...
 
                 commitErrorInjectedClient1.set(false);
                 stateTransitions1.clear();
                 streams1Beta.close();
-                waitForStateTransition(stateTransitions1, CLOSE_CRASHED);
+                waitForStateTransition(stateTransitions1, CLOSE);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: As above

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 22]
```diff
@@ -737,7 +729,7 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
                 errorInjectedClient2.set(false);
                 stateTransitions2.clear();
                 streams2AlphaTwo.close();
-                waitForStateTransition(stateTransitions2, CLOSE_CRASHED);
+                waitForStateTransition(stateTransitions2, CLOSE);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: As above

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: Not sure if I understand. Can you elaborate?

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java [hunk 23]
```diff
@@ -128,8 +128,8 @@ public void shouldShutdownThreadUsingOldHandler() throws InterruptedException {
             TestUtils.waitForCondition(() -> counter.get() == 1, "Handler was called 1st time");
             // should call the UncaughtExceptionHandler after rebalancing to another thread
             TestUtils.waitForCondition(() -> counter.get() == 2, DEFAULT_DURATION.toMillis(), "Handler was called 2nd time");
-            // the stream should now turn into ERROR state after 2 threads are dead
-            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.ERROR, DEFAULT_DURATION);
+            // there is no threads running but the client is still in running
+            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.RUNNING, DEFAULT_DURATION);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java:132]: Should we add a metric check to verify that the thread count is zero?

File: streams/src/test/java/org/apache/kafka/streams/integration/SuppressionIntegrationTest.java [hunk 24]
```diff
@@ -526,6 +526,6 @@ private static void produceSynchronously(final String topic, final List<KeyValue
 
     private static void verifyErrorShutdown(final KafkaStreams driver) throws InterruptedException {
         waitForCondition(() -> !driver.state().isRunningOrRebalancing(), DEFAULT_TIMEOUT, "Streams didn't shut down.");
-        assertThat(driver.state(), is(KafkaStreams.State.PENDING_SHUTDOWN));
+        waitForCondition(() -> driver.state() == KafkaStreams.State.ERROR, "finish shutdown");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/SuppressionIntegrationTest.java]: Nit: the error message is weird -> `Streams didn't transit to ERROR state`.

**Review by mjsax — COMMENTED** [review]
Made a pass over it.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 14 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: yay ascii art! good catch I forgot about that

When I build the docs locally it looks about right

image

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 15 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I don't think that we should return true. A user would then expect that that the state would then change to `NOT_RUNNING` and could be stuck waiting on that. Where if we return false in then they might retry and get stuck there.  

I don't know which is a better problem to have but I think that changing the meaning of this return value won't add any more clarity to the situation.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 16 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: It will. We don't strictly need to test it here as that is tested elsewhere but we can add it for clarity

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 7 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:537]: thats probably a good idea

File: streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java (hunk 8 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java:102]: It made it so that the new default was not used until we updated the Error transition as we are doing in this PR. :)

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 10 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: My thoughts where that we are calling close and so its is going to just be closed as the state transitions are cleared. This is the test that made me want you to review the PR I was worried that I might have broken the integrity of the tests as it is not behaving as I expected. However no other test was having this problem so I was not sure if I was understanding it correctly.

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: My theory (based on logs) is that the cleaner thread was sometimes activating and removing the segment before it should so the record that makes the thread crash was not being consumed by the recovery thread sometimes. I just changed the time stamps so that the cleaner thread would not find them old and if it did activate it would not clean them.

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java (hunk 23 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java:132]: We can but we should be able to see that with `counter` that the 2 threads already failed

**Review by wcarlson5 — COMMENTED** [review]
@mjsax I believe that I got to all your comments except for the `shouldUpgradeFromEosAlphaToEosBeta ` tests one

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 15 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Fair point. Might be good to get input from somebody else in addition. \cc @cadonna

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 16 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: I guess it would not hurt. Leave it up to you than.

File: streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java (hunk 8 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java:102]: Ah. Thanks for clarifying.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: From my understanding/ according to the new state transition diagram, if there is an error we go to `PENDING_ERROR` and if we call `close()` we transit `PENDING_ERROR -> ERROR`?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 25]
```diff
@@ -201,32 +201,33 @@
...
      *
      * </pre>
      * Note the following:
      * - RUNNING state will transit to REBALANCING if any of its threads is in PARTITION_REVOKED or PARTITIONS_ASSIGNED state
      * - REBALANCING state will transit to RUNNING if all of its threads are in RUNNING state
-     * - Any state except NOT_RUNNING can go to PENDING_SHUTDOWN (whenever close is called)
+     * - Any state except NOT_RUNNING, PENDING_ERROR or ERROR can go to PENDING_SHUTDOWN (whenever close is called)
      * - Of special importance: If the global stream thread dies, or all stream threads die (or both) then
-     *   the instance will be in the ERROR state. The user will need to close it.
+     *   the instance will be in the ERROR state. The user will not need to close it.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:222]: I thought we always need to call close? If an error happens, we call the handler, and if the handler return shutdown, we transit to `PENDING_ERROR`. On `close()` we transit from `PENDING_ERROR -> ERROR`?

Or do I have some misconception?

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: I see. You are talking about the broker side log cleaner thread -- not the local state directory cleaner thread. (windowed stores also use the concept of segments, not just topic-partitions).

However, it seems we want to extract this fix into a separate PR to be able to cherry-pick it to older branches? Seems to be unrelated to this PR.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 25 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:222]: the handler will call close, but the user will not need to. The `PENDING_ERROR` state is indicating the resources are closing before the transition to `ERROR` after which no more work will be done. We made it so the user can call close on `PENDING_ERROR` or `ERROR` but it will only log a warning

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 10 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: the call to `close()` will have no effect. The `PENDING_ERROR -> ERROR` transition is like the `PENDING_SHUTDOWN -> NOT_RUNNING ` where the transition from the `PENDING...` state is automatic

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: Maybe we will. It doesn't seems to be a problem for the other branches but something with how the test changes in this PR exposed it. This happened in the handler int tests too. I think as long as the other tests don't change it won't be a problem and if those tests change we can fix it then.

But if you think it should be fixed in the other branches anyways I'll trust your judgement.

**Review by wcarlson5 — COMMENTED** [review]
@mjsax I am still looking into the `shouldUpgradeFromEosAlphaToEosBeta` test. but not making a ton of progress I will try to carve out a little time tomorrow for it.

I think there was a bit of miss-understanding about what `PENDING_ERROR` means but I hope I cleared it up the comments below.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 25 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:222]: Thanks for clarifying!

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:541]: Thanks for clarifying.

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: I don't understand why other branches would not be affected, or to what extend this PR would expose a new issue? Can you elaborate? -- If this PR introduces an issue that requires this change, I agree that we don't need to port it to other branches -- but I don't see the connection. What do I miss?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 18 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Actually it could probably be just `PENDING_ERROR -> ERROR` because it could be in `RUNNING` or `REBALNCING` previously

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: It looks like this issue is fixed by another PR. https://github.com/apache/kafka/pull/9733#issuecomment-761242084 So I am closing this now

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 26]
```diff
@@ -113,14 +114,7 @@
...
             Collections.singletonList(
-                KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.ERROR)
-            )
-        );
-    private static final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> CLOSE_CRASHED =
-        Collections.unmodifiableList(
-            Arrays.asList(
-                KeyValue.pair(KafkaStreams.State.ERROR, KafkaStreams.State.PENDING_SHUTDOWN),
-                KeyValue.pair(KafkaStreams.State.PENDING_SHUTDOWN, KafkaStreams.State.NOT_RUNNING)
+                KeyValue.pair(State.PENDING_ERROR, State.ERROR)
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:117]: Using just `PENDING_ERROR -> ERROR` because the transition to `PENDING_ERROR` can be from multiple sources. Also that transition is already tested so this check implies it

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 15 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: As far as I can see, where the return value is used the javadoc says

```
true if all threads were successfully stopped, false if the timeout was reached.
```

Since all threads were successfully stopped, I would return `true`. We clearly document that `ERROR` is a terminal state, so I do not see why somebody should wait for `NOT_RUNNING` when the client is in `ERROR` and `close()` returns `true`.

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Alright I adjusted the close response to align with this.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 27]
```diff
@@ -978,7 +971,7 @@ private void waitForRunning(final List<KeyValue<KafkaStreams.State, KafkaStreams
         waitForCondition(
             () -> !observed.isEmpty() && observed.get(observed.size() - 1).value.equals(State.RUNNING),
             MAX_WAIT_TIME_MS,
-            () -> "Client did not startup on time. Observers transitions: " + observed
+            () -> "Client did not have the expected state transition on time. Observers transitions: " + observed
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Why this change? We do wait for `RUNNING`?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 28]
```diff
@@ -993,6 +986,17 @@ private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, Kafk
         );
     }
 
+    private void waitForStateTransitionContains(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: nit: fix indention

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 29]
```diff
@@ -993,6 +986,17 @@ private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, Kafk
...
+            throws Exception {
+
+        waitForCondition(
+            () -> observed.containsAll(expected),
+            MAX_WAIT_TIME_MS,
+            () -> "Client did not have the expected state transition on time. Observers transitions: " + observed
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:997]: Can we add the expected transitions, too? Easier to debug if the test fails.

File: streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java (hunk 12 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java]: The PR you liked seems to be unrelated to this test.

Still wondering if we should extract this change to a dedicated PR and cherry-pick to older branches? -- Or do we have a good explanation why older branches would not be affected?

**Review by mjsax — COMMENTED** [review]
Just some final nits. Overall LGTM.

**Comment by mjsax** [comment]
Btw: to what extend do we need to update the docs? We should at least add a section to `streams/upgrade_guide.html` to mention the change.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 27 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: We do wait for running, I was thinking of bringing it to match with the other methods below but that doesn't make it anymore useful so I will revet it.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 29 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:997]: sure that is fine

**Comment by wcarlson5** [comment]
I extracted the https://github.com/apache/kafka/pull/9720/files#r562203226 fix to https://github.com/apache/kafka/pull/9948

**Comment by wcarlson5** [comment]
@mjsax it looks like there is one unrelated failure `java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TopicExistsException: Topic 'primary.test-topic-2' already exists.
`

**Comment by glasser** [comment]
Am I correct in believing this change was released in Kafka 2.8.0? I can't find it in [the release notes](https://archive.apache.org/dist/kafka/2.8.0/RELEASE_NOTES.html). We just had a minor outage because we didn't realize when upgrading to v2.8 that some error handling/shutdown code in our Kafka Streams app couldn't assume that it only had to kick in when the state transitioned to `NOT_RUNNING` but also now has to handle `ERROR` too.
